pennylane/beta/tapes/tape.py

```diff
@@ -904,6 +909,15 @@ def execute_device(self, params, device):
             params (list[Any]): The quantum tape operation parameters. If not provided,
                 the current tape parameter values are used (via :meth:`~.get_parameters`).
         """
+        if self._state_measurement:
+            if not device.capabilities().get("returns_state"):
+                raise ValueError("The current device is not capable of returning the state")
+            if len(self.measurements) > 1:
+                raise ValueError(
+                    "The state cannot be returned in combination with other return " "types"
+                )
+            self._state_measurement._wires = device.wires
```

Comment:
This also feels kind of out of place, but I understand why it's been done. The `MeasurementProcess` was a way for us to 'offload' the measurement logic out of the QNode/tape, and encapsulate it in an object that everything can interpret and understand (the other exception so far has been `output_dim`. Oh!! I just had an idea. We should add `output_dim` to the measurement process, instead of defining it in the tape!!!!)

Ideally, it would be better if we could simply pass a `MeasurementProcess(State)` to the device, and the device just knows by convention to return the state on all wires. Especially since this is the convention on the user-level --- `return qml.state()` acts on all wires.

Reply:
Similar to in comment https://github.com/PennyLaneAI/pennylane/pull/818#discussion_r493553930, this section has been moved out to the device. I also realised that line 919 above actually isn't needed!